zygrader/ui/layers.py

```diff
@@ -0,0 +1,634 @@
+        self.__text = text
+        self.sort_index = 0
+        self.color = 0
+
+        self.__subrows: List[self.__class__] = []
+        self.__callback_fn = None
+        self._callback_args = None
+
+        self.__expanded = False
+
+        self.__toggle: Toggle = None
+        self.__radio: Radio = None
+
+    def __str__(self):
+        """Render a textual representation of this row."""
+        if self.__type == Row.TEXT:
+            return self.__text
+        if self.__type == Row.PARENT:
+            return f" {'v' if self.__expanded else '>'} " + self.__text
+        elif self.__type == Row.TOGGLE:
+            return f"[{'x' if self.__toggle.is_toggled() else ' '}] " + self.__text
+        elif self.__type == Row.RADIO:
+            return f"({'x' if self.__radio.is_toggled(self.__text) else ' '}) " + self.__text
+
+    def build_string_lines(self, lines, start, depth=0):
+        OFFSET = " " * (4 * depth)
+        for row in start.get_subrows():
+            lines.append((OFFSET + str(row), row.color, row.sort_index))
+            if row.get_type() == Row.PARENT and row.is_expanded():
+                self.build_string_lines(lines, row, depth + 1)
+
+    def __add_row(self, text: str, _type=TEXT):
+        row = Row(text, _type)
+        self.__subrows.append(row)
+        return row
+
+    def add_row_text(self, text: str, callback_fn=None, *args):
+        row = self.__add_row(text)
+        row.set_callback_fn(callback_fn, *args)
+        return row
+
+    def add_row_parent(self, text: str):
+        return self.__add_row(text, Row.PARENT)
+
+    def add_row_toggle(self, text: str, toggle: Toggle):
+        row = self.__add_row(text, Row.TOGGLE)
+        row.set_toggle_ob(toggle)
+        return row
+
+    def add_row_radio(self, text: str, radio: Radio):
+        row = self.__add_row(text, Row.RADIO)
+        row.set_radio_ob(radio)
+        return row
+
+    def get_type(self):
+        return self.__type
+
+    def get_subrows(self):
+        return self.__subrows
+
+    def is_expanded(self):
+        return self.__expanded
+
+    def set_callback_fn(self, callback_fn, *args):
+        self.__callback_fn = callback_fn
+        self._callback_args = args
+
+    def set_toggle_ob(self, toggle: Toggle):
+        self.__toggle = toggle
+
+    def set_radio_ob(self, radio: Radio):
+        self.__radio = radio
+
+    def set_row_text(self, text):
+        self.__text = text
+
+    def set_row_sort_index(self, sort_index):
+        self.sort_index = sort_index
+
+    def set_row_color(self, color):
+        self.color = color
+
+    def set_subrow_text(self, text, index):
+        self.__subrows[index].set_row_text(text)
+
+    def clear_rows(self):
+        self.__subrows.clear()
+
+    def do_action(self):
+        if self.__type == Row.TEXT and self.__callback_fn:
+            if self._callback_args:
+                self.__callback_fn(*self._callback_args)
+            else:
+                self.__callback_fn()
```

Comment:
If callback args is defaulted to `()`, this should be able to become
```suggestion
            self.__callback_fn(*self._callback_args)
```